File: haskell/p07.hs
-- P07 Flatten a nested list structure.

module P07 where

data NestedList a = Value a | List [NestedList a]

-- Recursion
f1 :: NestedList a -> [a]
f1 (Value a) = [a]
f1 (List []) = []
f1 (List (x:xs)) = (f1 x) ++ f1(List xs)

-- Using concatMap
f2 :: NestedList a -> [a]
f2 (Value a) = [a]
f2 (List xs) = concatMap f2 xs

File: scala/P07.scala
// P07 Flatten a nested list structure.

// flatMap with recursion, unchecked
def f1(list: List[Any]): List[Any] = list flatMap {
  case xs: List[_] => f1(xs)
  case v => List(v)
}

// With custom data
abstract sealed class NestedList[T]
case class Value[T](v: T) extends NestedList[T]
case class Sequence[T](list: List[NestedList[T]]) extends NestedList[T]

def f2[T]: NestedList[T] => List[T] = {
  case Value(v) => List(v)
  case Sequence(list) => list.flatMap(f2)
}